auth: implement vector store authorization

smoczy123 opened the pull request:
This patch implements the changes required by the Vector Store authorization, as described in https://scylladb.atlassian.net/wiki/spaces/RND/pages/107085899/Vector+Store+Authentication+And+Authorization+To+ScyllaDB, that is:

- adding a new permission VECTOR_SEARCH_INDEXING, grantable only on ALL KEYSPACES
- allowing users with that permission to perform SELECT queries, but only on tables with a vector index
- increasing the number of scheduling groups by one to allow users to create a service level for a vector store user
- adjusting the tests and documentation

These changes are needed, as the vector indexes are managed by the external service, Vector Store, which needs to read the tables to create the indexes in its memory. We would like to limit the privileges of that service to a minimum to maintain the principle of least privilege, therefore a new permission, one that allows the SELECTs conditional on the existence of a vector_index on the table.

Fixes: VECTOR-201
Fixes: https://github.com/scylladb/scylladb/issues/26804

Backport reasoning: 
Backport to 2025.4 required as this can make upgrading clusters more difficult if we add it in 2026.1. As for now Scylla Cloud requires version 2025.4 to enable vector search and permission is set by orchestrator so there is no chance that someone will try to add this permission during upgrade. In 2026.1 it will be more difficult.

andrzej-jackowski-scylladb commented on auth/resource.cc:
```diff
@@ -18,6 +18,7 @@
 #include <boost/algorithm/string/split.hpp>
 #include <boost/algorithm/string/classification.hpp>
 
+#include "auth/permission.hh"
```
Is it needed? It seems to be already included in "auth/resource.hh"

nuivall commented on auth/resource.cc:
```diff
@@ -41,6 +42,8 @@ static const std::unordered_map<resource_kind, std::size_t> max_parts{
 
 static permission_set applicable_permissions(const data_resource_view& dv) {
+    
+    // table
```
I think this comment doesn't provide much information (if any)

nuivall commented on auth/resource.cc:
```diff
@@ -50,13 +53,27 @@ static permission_set applicable_permissions(const data_resource_view& dv) {
             permission::AUTHORIZE>();
+    }
+
+    // All keyspaces
```
We need more information here, why this split dv.keyspace vs all keyspaces.

Also you can probably use add() to permission_set to keep some common code

knowack1 commented on auth/permission.hh:
```diff
@@ -40,6 +40,9 @@ enum class permission {
     EXECUTE,
+
+    // Vector Store service permission
+    VECTOR_SEARCH_INDEXING // allows selecting from tables with vector indexes
```
Consider: Keep comments style: "// required ..."

knowack1 commented on auth/permission.hh:
```diff
@@ -40,6 +40,9 @@ enum class permission {
     // function/aggregate/procedure calls
     EXECUTE,
+
+    // Vector Store service permission
```
Consider: this permission fits to `// data access` list. Maybe Could be places in this group.

andrzej-jackowski-scylladb commented on test/boost/cql_auth_query_test.cc:
```diff
@@ -6,6 +6,8 @@
  * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
  */
 
+#include "test/lib/exception_utils.hh"
```
Please move the include after `#include "test/lib/cql_assertions.hh"`

nyh commented on auth/resource.cc:
```diff
@@ -43,36 +43,29 @@ static const std::unordered_map<resource_kind, std::size_t> max_parts{
-    // keyspace
+    set.add(permission_set::of<permission::CREATE>());
+
+    // Permissions applicable to a given keyspace - we cannot grant CREATE permission on an already created table
```
This comment seems misplaced. It explains why 10 lines higher there is no "CREATE" listed for permissions for a table.

Why did you need to change this code at all - wasn't it already clear enough, perhaps even clearer?

Moreover, you are removing comments that you just add in the previous patch :-)

smoczy123 replied:
The change is needed as we decided to make the new permission grantable only on ALL KEYSPACES. I've rewritten it a bit, it should be clearer now.

nyh commented on auth/resource.hh:
```diff
@@ -100,13 +100,16 @@ class resource final {
 
     utils::small_vector<sstring, 3> _parts;
 
+    bool is_vector_indexed_table = false;
```
You need a comment explaining what "is_vector_indexed_table" means. I couldn't understand it at all until I found service::client_state::has_column_family_access() that uses it.

I don't like this super-specific vector-search modification to the very general resource model of RBAC, but I guess that ship has already sailed.

smoczy123 replied:
I agree that this is ugly, however as the new behavior of auth depends on the schema we have to pass that new information somehow. I couldn't find a better way. I've added an explanation comment.

knowack1 replied:
This generic code is already aware of vector search through the new permissions::VECTOR_SEARCH_INDEXING permission. Therefore, the new is_vector_indexed_table boolean doesn't change much in this context.

However, it's worth noting that this new permission fundamentally changes the design of access validation.
Until now, validating access only required two pieces of information:
1. Access resource (from the CQL query)
2. User/role permissions (held by the client state)

With this change, validation now requires a third element:
1. Access resource (from the CQL query)
2. User/role permissions (held by the client state)
3. **Schema** (to verify if the resource has a specific index, in this case, a vector index)

To me, this suggests we should take a step back and reconsider the built-in role approach instead.

nyh commented on test/cluster/auth_cluster/test_raft_service_levels.py:
```diff
@@ -433,7 +433,7 @@ async def test_service_levels_over_limit(manager: ManagerClient):
     hosts = await wait_for_cql_and_get_hosts(cql, [srv], time.time() + 60)
 
-    SL_LIMIT = 7
+    SL_LIMIT = 8
```
I'm lost how the 8s in this patch and the following patch relates to the 7s in a different patch I saw you sent.

smoczy123 replied:
I'm not sure what you mean, could you provide a reference to the other patch?

smoczy123 commented on auth/resource.cc:
```diff
@@ -119,6 +119,12 @@ resource::resource(data_resource_t, std::string_view keyspace, std::string_view
+resource::resource(data_resource_t, std::string_view keyspace, std::string_view table, bool is_vector_indexed) : resource(resource_kind::data) {
+    _parts.emplace_back(keyspace);
+    _parts.emplace_back(table);
+    is_vector_indexed_table = is_vector_indexed;
```
We would have to remove the delegating constructor from the initializer list as you can't mix them.

smoczy123 commented:
Changelog:

- Cleared up the resource constructor

smoczy123 commented on service/client_state.cc:
```diff
@@ -224,6 +224,13 @@ future<> service::client_state::has_access(const sstring& ks, auth::command_desc
+    if (cmd.resource.kind() == auth::resource_kind::data
+            && cmd.permission == auth::permission::SELECT
+            && cmd.resource.is_vector_indexed_table()
+            && co_await check_has_permission({auth::permission::VECTOR_SEARCH_INDEXING, cmd.resource, cmd.type_})) {
```
Hmmm, it looks like that would be quite difficult as a lot of code depends on the command having one permission, I'll try to think of something

nuivall commented on service/client_state.cc:
```diff
@@ -244,6 +253,23 @@ future<bool> service::client_state::check_has_permission(auth::command_desc cmd)
     co_return false;
 }
 
+future<bool> service::client_state::check_has_permissions(auth::command_desc cmd, const auth::permission_set& perms) const {
```
I see 3 problems here:
- name of the function doesn't hint if this is ALL or ANY check.
- code duplication could be avoided
- it's not ok that in some cases we validate on cmd.permission and other on perms arg.

nuivall replied:
Or create different command type with set, and overload checking permissions function, or template it. There are many solutions but we have 2 constrains:
- no perf degradation
- code is not confusing or duplicated

nyh commented:
@smoczy123 there are multiple failing tests that seem to be related to what you're changing (authentication), so please take a look.

smoczy123 commented on service/client_state.hh:
```diff
@@ -332,7 +332,7 @@ public:
     future<> has_keyspace_access(const sstring&, auth::permission) const;
     future<> has_column_family_access(const sstring&, const sstring&, auth::permission,
-                                      auth::command_desc::type = auth::command_desc::type::OTHER) const;
+                                      auth::command_desc::type = auth::command_desc::type::OTHER, bool is_vector_indexed = false) const;
```
Unfortunately we cannot differentiate between false and unknown, as when accessing system resources we check if they are contained in some set. If we differentiate those two states, this check fails. I'm not sure what the best solution here would be, maybe some rename of this field?

nuivall commented on service/client_state.cc:
```diff
@@ -224,38 +224,77 @@ future<> service::client_state::has_access(const sstring& ks, auth::command_desc
+template future<bool> service::client_state::check_has_permission<auth::command_desc_with_permission_set>(auth::command_desc_with_permission_set) const;
 
-future<> service::client_state::ensure_has_permission(auth::command_desc cmd) const {
+template <typename Cmd>
```
maybe worth put a constraint that it's either of those two cmd types we have? not a must have though

nyh commented on docs/features/workload-prioritization.rst:
```diff
@@ -425,7 +425,7 @@ In order for workload prioritization to take effect, application users need to b
 Limits
 ======
-ScyllaDB is limited to 8 service levels, including the default one; this means you can create up to 7 service levels.
+ScyllaDB is limited to 9 service levels, including the default one; this means you can create up to 8 service levels.
```
If we keep the service level change in this PR (I don't know if we should...) please fold this specific change into the earlier patch which did the other 8->9 changes.

nyh reviewed (commented):
Looks good to me, but a comment by @mykaul in another issue reminded me - do we need a feature flag?
The problem is what happens on a partially upgraded cluster, where a user connected to a new node asks to grant a "VECTOR_SEARCH_INDEXING" permission. How does this information reach new nodes and how do they handle this?

smoczy123 commented on test/boost/cql_auth_query_test.cc:
```diff
@@ -6,6 +6,7 @@
  * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
  */
 
+#include <boost/test/tools/old/interface.hpp>
```
Not sure how it got there, removed.

nuivall commented on service/client_state.cc:
```diff
@@ -224,38 +224,77 @@ future<> service::client_state::has_access(const sstring& ks, auth::command_desc
+    return permissions.intersects(cmd.permission);
+}
+
+static bool intersects_permissions(const auth::permission_set& permissions, const auth::command_desc& cmd) {
```
something not right, one intersects_ func calls permissions.intersects and the other permissions.contains ?

nuivall replied:
but the naming is super confusing, no?

nuivall commented on service/client_state.cc:
```diff
@@ -224,38 +224,77 @@ future<> service::client_state::has_access(const sstring& ks, auth::command_desc
+sstring service::client_state::generate_authorization_error_msg(const auth::command_desc_with_permission_set& cmd) const {
+    std::ostringstream os;
+    for (auto p : auth::permissions::to_strings(cmd.permission)) {
+        if (os.tellp() > 0) {
```
maybe fmt::join can be used instead?

nuivall commented on service/client_state.cc:
```diff
@@ -93,8 +93,8 @@ future<> service::client_state::has_function_access(const sstring& ks, const sst
-                const sstring& cf, auth::permission p, auth::command_desc::type t) const {
-    auto r = auth::make_data_resource(ks, cf);
+                const sstring& cf, auth::permission p, auth::command_desc::type t, bool is_vector_indexed) const {
+    auto r = auth::make_data_resource(ks, cf, is_vector_indexed);
```
Perhaps the resource way is marginally better, because it's more clear that generic object can represent specific cases i.e. when is_vector_indexed is unknown by using optional. And is_vector_indexed = false in a function call is less clear, but it can also be made optional, still I'd prefer documenting it in an object than function signature. But I can live with both.

smoczy123 commented:
Changelog: 

- rebased onto master
- changed stream error generation to fmt::join
- replaced the bool argument of the has_access function with optional<bool> (doesn't hurt us now as it is not in resource)

smoczy123 commented:
@scylladbbot trigger-ci

nyh commented on service/client_state.cc:
```diff
@@ -93,9 +93,9 @@ future<> service::client_state::has_function_access(const sstring& ks, const sst
 
 future<> service::client_state::has_column_family_access(const sstring& ks,
-                const sstring& cf, auth::permission p, auth::command_desc::type t) const {
+                const sstring& cf, auth::permission p, auth::command_desc::type t, std::optional<bool> is_vector_indexed) const {
```
How did we reach a situation where this flag needs to be an optional<bool> - can't it just be a bool?
More on this below.

nyh commented on service/client_state.hh:
```diff
@@ -332,7 +332,7 @@ public:
     future<> has_keyspace_access(const sstring&, auth::permission) const;
     future<> has_column_family_access(const sstring&, const sstring&, auth::permission,
-                                      auth::command_desc::type = auth::command_desc::type::OTHER) const;
+                                      auth::command_desc::type = auth::command_desc::type::OTHER, std::optional<bool> is_vector_indexed = std::nullopt) const;
```
I'm starting to think that you added the extra later of optional because of Python guide books?
You can just have "bool is_vector_index = false" here as the default argument, there is no need for the default argument to be a nullopt.

smoczy123 replied:
I've made this argument into an optional due to a discussion in this pull request about a situation when the table is actually vector indexed but we don't care about it. In that case we were passing false to the function which was not reflecting the actual state of the table. We were considering renaming this argument to make this more clear, however no good name was decided on and we settled on making this into an optional, the situation where we don't care about the vector indexing represented by nullopt.
